### PR TITLE
Add cancelLastOperation() API

### PR DESCRIPTION
**Pull request by tikurahul**
This adds the `cancelLastOperation()` API. I added one extra test on the `Provider` frame to prove that this works. If this approach looks good, I can add more tests before I submit the change.

### CODE REVIEW

File: ts/api/api.ts [hunk 1]
```diff
@@ -325,6 +331,14 @@ class OpenYoloApiImpl implements OpenYoloApi {
         this.areTimeoutsDisabled ? undefined : DEFAULT_TIMEOUTS.proxyLogin;
     return request.dispatch(credential, timeoutMs);
   }
+
+  async cancelLastOperation() {
+    let request =
+        new CancelLastOperationRequest(this.frameManager, this.channel);
+    const timeoutMs =
+        this.areTimeoutsDisabled ? undefined : DEFAULT_TIMEOUTS.credentialSave;
```
Comment (TMSCH) [inline ts/api/api.ts]: Use a custom DEFAULT_TIMEOUTS.cancelLastOperation value.

File: ts/api/api.ts [hunk 2]
```diff
@@ -325,6 +331,14 @@ class OpenYoloApiImpl implements OpenYoloApi {
         this.areTimeoutsDisabled ? undefined : DEFAULT_TIMEOUTS.proxyLogin;
     return request.dispatch(credential, timeoutMs);
   }
+
+  async cancelLastOperation() {
```
Comment (TMSCH) [inline ts/api/api.ts]: You should do like all the other requests, i.e. have one using the browser (i.e. in this case, raise an error probably as you cannot cancel) and one that uses the IFrame.

File: ts/spi/provider_frame.ts [hunk 3]
```diff
@@ -544,4 +571,10 @@ export class ProviderFrame {
 
     return score;
   }
+
+  private race<T>(value: T, producer: Promise<T>): Promise<T> {
```
Comment (iainmcgin) [inline ts/spi/provider_frame.ts]: I would call this method `cancellable` or similar, and rename value to cancelValue, probably move it to the second argument position too.

File: ts/spi/provider_frame.ts [hunk 4]
```diff
@@ -344,6 +354,22 @@ export class ProviderFrame {
     return this.handleUnimplementedRequest(id, msg.RPC_MESSAGE_TYPES.proxy);
   }
 
+  private async handleCancelLastOperation(id: string) {
+    if (!this.requestInProgress || this.cancellable === null) {
+      // no request in progress
+      console.warn('No pending request to cancel.');
+      this.clientChannel.send(msg.errorMessage(
+          id, OpenYoloError.requestFailed('No request in progress.')));
```
Comment (iainmcgin) [inline ts/spi/provider_frame.ts]: I would perhaps just treat this as a no-op; I think your response here will result in the cancel promise being rejected, which could have consequences for the client code.

File: ts/protocol/utils.ts [hunk 5]
```diff
@@ -192,3 +192,20 @@ export function timeoutPromise<T>(error: Error, timeoutMs: number): Promise<T> {
...
+/**
+ * Is a cancellable PromiseResolver.
+ */
+export class CancellablePromise<T> extends PromiseResolver<T> {
+  constructor(private readonly value: T) {
+    super();
+  }
+
+  cancel() {
+    this.resolve(this.value);
```
Comment (TMSCH) [inline ts/protocol/utils.ts]: Or reject? Wouldn't it make more sense?

File: ts/spi/provider_frame.ts [hunk 6]
```diff
@@ -344,6 +354,22 @@ export class ProviderFrame {
...
+    } else {
+      try {
+        this.cancellable.cancel();
+      } finally {
+        this.dispose();
```
Comment (iainmcgin) [inline ts/spi/provider_frame.ts]: do you need to dispose the frame?

File: ts/spi/provider_frame.ts [hunk 7]
```diff
@@ -33,12 +34,15 @@ export class ProviderFrame {
...
   private requestInProgress = false;
+  // represents a potential cancellable operation
+  private cancellable: CancellablePromise<{}>|null = null;
 
   private closeListener: EventListener;
   private window: WindowLike;
 
   private proxyLoginCredential: Credential|null = null;
 
+
```
Comment (TMSCH) [inline ts/spi/provider_frame.ts]: Remove the extra line.

File: ts/spi/provider_frame_test.ts [hunk 8]
```diff
@@ -300,6 +300,40 @@ describe('ProviderFrame', () => {
...
+                });
+
+        let credentialPromise = new PromiseResolver<void>();
+        let cancellationPromise = new PromiseResolver<void>();
+
+        clientChannel.listen(msg.RPC_MESSAGE_TYPES.credential, (data) => {
+          expectMessageContents(
+              data,
+              msg.credentialResultMessage(
+                  requestId, undefined as any as Credential));
```
Comment (iainmcgin) [inline ts/spi/provider_frame_test.ts]: `undefined as any as Credential` - wat

File: ts/spi/provider_frame.ts [hunk 9]
```diff
@@ -344,6 +354,22 @@ export class ProviderFrame {
     return this.handleUnimplementedRequest(id, msg.RPC_MESSAGE_TYPES.proxy);
   }
 
+  private async handleCancelLastOperation(id: string) {
+    if (!this.requestInProgress || this.cancellable === null) {
+      // no request in progress
+      console.warn('No pending request to cancel.');
+      this.clientChannel.send(msg.errorMessage(
```
Comment (TMSCH) [inline ts/spi/provider_frame.ts]: If this goes here, then two messages are sent (the requestFailed and the cancelLastOperation). Is that intended?

File: ts/spi/provider_frame.ts [hunk 10]
```diff
@@ -221,7 +229,7 @@ export class ProviderFrame {
       options: CredentialHintOptions) {
     console.info('Handling hint request');
 
-    let hints = await this.getHints(options);
+    let hints = await this.race([], this.getHints(options));
```
Comment (TMSCH) [inline ts/spi/provider_frame.ts]: You cannot only race here. The whole function has to be cancellable (as it is more likely that the need to cancel is after the UI is displayed, i.e. below when the interaction provider is called).

Didn't we discuss to race the cancellable promise at a higher level? (maybe in the `registerListeners`?)

File: ts/spi/provider_frame.ts (hunk 6 above)
Comment (tikurahul) [inline ts/spi/provider_frame.ts]: Yes, that is the only way to dispose of any pending UI right now. @TMSCH has been talking about adding a method on interaction provider which dismisses (and animates) all UI away.

File: ts/spi/provider_frame_test.ts (hunk 8 above)
Comment (tikurahul) [inline ts/spi/provider_frame_test.ts]: 😄 The `expectMessageContents(...)` method expects an instance of type `Credential`. This is the only way for it to understand that the result of a cancelled action will be an undefined.

File: ts/protocol/utils.ts (hunk 5 above)
Comment (tikurahul) [inline ts/protocol/utils.ts]: I think we don't reject `Promise`'s anywhere at the moment. Also adding result codes, will make this cleaner.

File: ts/spi/provider_frame.ts (hunk 10 above)
Comment (tikurahul) [inline ts/spi/provider_frame.ts]: I am not sure I understand the concern. When you call `await`, subsequent interactions with the result of the await are being passed on to you via a generator's `next(...)` method. So adding it to the top level promise will cascade to all future promises underneath the `await` call. 

Are you talking about something else ?

File: ts/spi/provider_frame.ts (hunk 9 above)
Comment (tikurahul) [inline ts/spi/provider_frame.ts]: No. I am going to treat the no pending operation case as a `no-op`.

File: ts/api/api.ts (hunk 2 above)
Comment (tikurahul) [inline ts/api/api.ts]: Are you talking about the tests or something else ? The wrapped browser APIs are in `navigator_credentials.ts` which is already been implemented.

File: ts/spi/provider_frame.ts (hunk 10 above)
Comment (tikurahul) [inline ts/spi/provider_frame.ts]: Never mind i think i got it. 😃

File: ts/protocol/utils.ts (hunk 5 above)
Comment (TMSCH) [inline ts/protocol/utils.ts]: We do. The TimeoutPromiseResolver rejects on timeout: https://github.com/openid/OpenYOLO-Web/blob/master/ts/protocol/utils.ts#L183 - or where you talking about something different? (we also reject promise very often across the codebase)

When you mention result codes, are you saying *error codes*? The results are determined by the type of the Promise (i.e. `Credential|null`, `boolean`, etc.), and cannot contain a *code*. The errors, however, would benefit from having codes.

File: ts/protocol/errors.ts [hunk 11]
```diff
@@ -64,6 +65,13 @@ export class OpenYoloError {
         {code: ERROR_TYPES.canceled, message: 'User canceled'});
   }
 
+  static cancelledLastOperation() {
+    return OpenYoloError.createError({
+      code: ERROR_TYPES.canceledLastOperation,
```
Comment (iainmcgin) [inline ts/protocol/errors.ts]: I would make this `clientCanceled`

Comment (TMSCH) [inline ts/protocol/errors.ts]: And be consistent with canceled/cancelled ;)

File: ts/api/cancel_last_operation_request.ts [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+ * limitations under the License.
+ */
+
+import {cancelLastOperationMessage, RPC_MESSAGE_TYPES} from '../protocol/rpc_messages';
+import {BaseRequest} from './base_request';
+
+export class CancelLastOperationRequest extends BaseRequest<void, void> {
+  dispatchInternal() {
+    this.registerHandler(
+        RPC_MESSAGE_TYPES.cancelLastOperationResult, this.handleResult);
```
Comment (TMSCH) [inline ts/api/cancel_last_operation_request.ts]: Add tests. Calling this.handleResult like that will lose the reference to this.

File: ts/api/api.ts [hunk 13]
```diff
@@ -325,6 +332,23 @@ class OpenYoloApiImpl implements OpenYoloApi {
         this.areTimeoutsDisabled ? undefined : DEFAULT_TIMEOUTS.proxyLogin;
     return request.dispatch(credential, timeoutMs);
   }
+
+  private async cancelLastOperationUsingBrowser() {
+    return Promise.reject('not implemented');
+  }
+
+  async cancelLastOperation() {
```
Comment (TMSCH) [inline ts/api/api.ts]: Add tests.

File: ts/protocol/utils.ts [hunk 14]
```diff
@@ -192,3 +192,21 @@ export function timeoutPromise<T>(error: Error, timeoutMs: number): Promise<T> {
...
 }
+
+/**
+ * Is a cancellable PromiseResolver.
+ */
+export class CancellablePromise<T> extends PromiseResolver<T> {
+  private static CANCELLED = 'CancellablePromise { Operation Cancelled }';
+
+  // the error returned when a cancellable promise is cancelled.
+  static CANCELLED_ERROR: Error&never = new Error(CancellablePromise.CANCELLED);
```
Comment (TMSCH) [inline ts/protocol/utils.ts]: I went through the discussion you had yesterday, but I thought the `never` was meant to be used as the type of CancellablePromise's template. Why using &never here?

File: ts/spi/provider_frame_test.ts [hunk 15]
```diff
@@ -300,6 +300,47 @@ describe('ProviderFrame', () => {
                msg.retrieveMessage(requestId, passwordOnlyRequest));
          });
 
+      it('should cancel a credential request', async function(done) {
```
Comment (TMSCH) [inline ts/spi/provider_frame_test.ts]: Add tests to make sure the proper messages are sent back to the client.

File: .nvmrc [hunk 16]
```diff
@@ -0,0 +1 @@
+stable
```
Comment (TMSCH) [inline .nvmrc]: Do we need this file?

Comment (tikurahul) [inline .nvmrc]: Yes, its convenient when we use NVM to provision an installation of node.

File: ts/spi/provider_frame_test.ts (hunk 15 above)
Comment (tikurahul) [inline ts/spi/provider_frame_test.ts]: Done.

File: ts/protocol/utils.ts (hunk 14 above)
Comment (tikurahul) [inline ts/protocol/utils.ts]: Removed. This was my misunderstanding of the purpose of a `never` type.

File: ts/api/api.ts (hunk 13 above)
Comment (tikurahul) [inline ts/api/api.ts]: Yes. My next update will make sure i test this via a demo provider like @iainmcgin suggested.

File: ts/api/cancel_last_operation_request.ts (hunk 12 above)
Comment (tikurahul) [inline ts/api/cancel_last_operation_request.ts]: Done.

File: ts/protocol/errors.ts (hunk 11 above)
Comment (tikurahul) [inline ts/protocol/errors.ts]: Done

File: ts/api/api.ts (hunk 13 above)
Comment (tikurahul) [inline ts/api/api.ts]: Tested it end to end with the demo provider. Everything works. :gear:

File: ts/api/api_test.ts [hunk 17]
```diff
@@ -70,6 +71,19 @@ describe('OpenYolo API', () => {
         });
       });
 
+      describe('cancelLastRequest', () => {
```
Comment (TMSCH) [inline ts/api/api_test.ts]: There's also the equivalent test without timeout to add further down in the file, for completeness.

Comment (tikurahul) [inline ts/api/api_test.ts]: Done

**Comment by tikurahul** [comment]
Thanks @TMSCH !

**Pull request merged**